File: packages/go-sdk/oauth_provider_client.go
package authgateway

import (
	"context"
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"errors"
	"fmt"
	"net/http"
	"net/url"
	"strings"
	"sync"
	"time"

	"github.com/smilemakc/auth-gateway/packages/go-sdk/models"
)

// OAuthProviderClient implements an OAuth 2.0 / OIDC client for applications
// that want to use Auth Gateway as their identity provider.
//
// Use Cases:
//   - Your web/mobile app wants to authenticate users via Auth Gateway
//   - Your service needs to integrate with Auth Gateway as an OAuth provider
//   - You're building a CLI tool or IoT device that needs Auth Gateway login
//
// Supported Flows:
//   - Authorization Code Flow with PKCE (web/mobile apps)
//   - Device Authorization Flow (TVs, CLI tools, IoT)
//   - Client Credentials Flow (service-to-service)
//   - Refresh Token Flow
//
// Security Features:
//   - PKCE (Proof Key for Code Exchange) enabled by default
//   - State parameter for CSRF protection
//   - Nonce for ID token validation
//   - Automatic OIDC discovery
//
// Example:
//
//	client := authgateway.NewOAuthProviderClient(authgateway.OAuthProviderConfig{
//	    Issuer:       "https://auth.example.com",
//	    ClientID:     "your-client-id",
//	    ClientSecret: "your-client-secret",
//	    RedirectURI:  "https://yourapp.com/callback",
//	    Scopes:       []string{"openid", "profile", "email"},
//	})
//
//	authURL, _ := client.GetAuthorizationURL(ctx, nil)
//	// Redirect user to authURL.URL
//
//	tokens, _ := client.ExchangeCode(ctx, code, authURL.CodeVerifier)
//	userInfo, _ := client.GetUserInfo(ctx, tokens.AccessToken)
//
// For complete documentation, see OAUTH_PROVIDER.md

var (
	ErrAuthorizationPending = errors.New("authorization_pending")
	ErrSlowDown             = errors.New("slow_down")
	ErrAccessDenied         = errors.New("access_denied")
	ErrExpiredToken         = errors.New("expired_token")
)

type OAuthProviderConfig struct {
	Issuer       string
	ClientID     string
	ClientSecret string
	RedirectURI  string
	Scopes       []string
	UsePKCE      bool
	HTTPClient   *http.Client

	// Headers contains custom headers to include in every request.
	// Common headers: X-Application-ID, X-Client-Name, etc.
	Headers map[string]string
}

type OAuthProviderClient struct {
	config     OAuthProviderConfig
	httpClient *http.Client

	discovery     *models.OIDCDiscovery
	discoveryOnce sync.Once

	jwks   *models.JWKS
	jwksMu sync.RWMutex

	// Custom headers to include in every request
	headers   map[string]string
	headersMu sync.RWMutex
}

func NewOAuthProviderClient(config OAuthProviderConfig) *OAuthProviderClient {
	if config.HTTPClient == nil {
		config.HTTPClient = &http.Client{Timeout: 30 * time.Second}
	}
	if config.Scopes == nil {
		config.Scopes = []string{"openid"}
	}
	if !config.UsePKCE {
		config.UsePKCE = true
	}

	return &OAuthProviderClient{
		config:     config,
		httpClient: config.HTTPClient,
		headers:    config.Headers,
	}
}

// SetHeader sets a custom header to be included in all requests.
func (c *OAuthProviderClient) SetHeader(key, value string) {
	c.headersMu.Lock()
	defer c.headersMu.Unlock()

	if c.headers == nil {
		c.headers = make(map[string]string)
	}
	c.headers[key] = value
}

// SetHeaders sets multiple custom headers to be included in all requests.
func (c *OAuthProviderClient) SetHeaders(headers map[string]string) {
	c.headersMu.Lock()
	defer c.headersMu.Unlock()

	if c.headers == nil {
		c.headers = make(map[string]string)
	}
	for k, v := range headers {
		c.headers[k] = v
	}
}

// RemoveHeader removes a custom header.
func (c *OAuthProviderClient) RemoveHeader(key string) {
	c.headersMu.Lock()
	defer c.headersMu.Unlock()

	delete(c.headers, key)
}

// GetHeaders returns a copy of the current custom headers.
func (c *OAuthProviderClient) GetHeaders() map[string]string {
	c.headersMu.RLock()
	defer c.headersMu.RUnlock()

	result := make(map[string]string)
	for k, v := range c.headers {
		result[k] = v
	}
	return result
}

// SetApplicationID sets the X-Application-ID header for multi-tenant support.
func (c *OAuthProviderClient) SetApplicationID(appID string) {
	c.SetHeader("X-Application-ID", appID)
}

// SetClientName sets the X-Client-Name header for client identification.
func (c *OAuthProviderClient) SetClientName(name string) {
	c.SetHeader("X-Client-Name", name)
}

// applyHeaders applies custom headers to an HTTP request.
func (c *OAuthProviderClient) applyHeaders(req *http.Request) {
	c.headersMu.RLock()
	defer c.headersMu.RUnlock()

	for key, value := range c.headers {
		req.Header.Set(key, value)
	}
}

func (c *OAuthProviderClient) GetDiscovery(ctx context.Context) (*models.OIDCDiscovery, error) {
	var err error
	c.discoveryOnce.Do(func() {
		url := c.config.Issuer + "/.well-known/openid-configuration"
		req, reqErr := http.NewRequestWithContext(ctx, "GET", url, nil)
		if reqErr != nil {
			err = reqErr
			return
		}
		c.applyHeaders(req)

		resp, respErr := c.httpClient.Do(req)
		if respErr != nil {
			err = respErr
			return
		}
		defer resp.Body.Close()

		if resp.StatusCode != http.StatusOK {
			err = fmt.Errorf("discovery endpoint returned status %d", resp.StatusCode)
			return
		}

		c.discovery = &models.OIDCDiscovery{}
		err = json.NewDecoder(resp.Body).Decode(c.discovery)
	})

	if err != nil {
		return nil, err
	}
	return c.discovery, nil
}

func (c *OAuthProviderClient) GetJWKS(ctx context.Context) (*models.JWKS, error) {
	c.jwksMu.RLock()
	if c.jwks != nil {
		c.jwksMu.RUnlock()
		return c.jwks, nil
	}
	c.jwksMu.RUnlock()

	discovery, err := c.GetDiscovery(ctx)
	if err != nil {
		return nil, err
	}

	req, err := http.NewRequestWithContext(ctx, "GET", discovery.JwksURI, nil)
	if err != nil {
		return nil, err
	}
	c.applyHeaders(req)

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("JWKS endpoint returned status %d", resp.StatusCode)
	}

	c.jwksMu.Lock()
	defer c.jwksMu.Unlock()

	c.jwks = &models.JWKS{}
	if err := json.NewDecoder(resp.Body).Decode(c.jwks); err != nil {
		return nil, err
	}

	return c.jwks, nil
}

type AuthorizationURLResult struct {
	URL          string
	State        string
	Nonce        string
	CodeVerifier string
}

type AuthorizationURLOptions struct {
	Scope     string
	State     string
	Nonce     string
	Prompt    string
	LoginHint string
}

func (c *OAuthProviderClient) GetAuthorizationURL(ctx context.Context, opts *AuthorizationURLOptions) (*AuthorizationURLResult, error) {
	discovery, err := c.GetDiscovery(ctx)
	if err != nil {
		return nil, err
	}

	if opts == nil {
		opts = &AuthorizationURLOptions{}
	}

	state := opts.State
	if state == "" {
		state = generateRandomString(32)
	}

	nonce := opts.Nonce
	if nonce == "" {
		nonce = generateRandomString(32)
	}

	scope := opts.Scope
	if scope == "" {
		scope = strings.Join(c.config.Scopes, " ")
	}

	params := url.Values{
		"response_type": {"code"},
		"client_id":     {c.config.ClientID},
		"redirect_uri":  {c.config.RedirectURI},
		"scope":         {scope},
		"state":         {state},
		"nonce":         {nonce},
	}

	if opts.Prompt != "" {
		params.Set("prompt", opts.Prompt)
	}
	if opts.LoginHint != "" {
		params.Set("login_hint", opts.LoginHint)
	}

	result := &AuthorizationURLResult{
		State: state,
		Nonce: nonce,
	}

	if c.config.UsePKCE {
		pkce, err := GeneratePKCE()
		if err != nil {
			return nil, err
		}
		result.CodeVerifier = pkce.CodeVerifier
		params.Set("code_challenge", pkce.CodeChallenge)
		params.Set("code_challenge_method", pkce.CodeChallengeMethod)
	}

	result.URL = discovery.AuthorizationEndpoint + "?" + params.Encode()
	return result, nil
}

func (c *OAuthProviderClient) ExchangeCode(ctx context.Context, code, codeVerifier string) (*models.OAuthTokenResponse, error) {
	discovery, err := c.GetDiscovery(ctx)
	if err != nil {
		return nil, err
	}

	params := url.Values{
		"grant_type":   {"authorization_code"},
		"code":         {code},
		"redirect_uri": {c.config.RedirectURI},
		"client_id":    {c.config.ClientID},
	}

	if c.config.ClientSecret != "" {
		params.Set("client_secret", c.config.ClientSecret)
	}

	if codeVerifier != "" {
		params.Set("code_verifier", codeVerifier)
	}

	return c.tokenRequest(ctx, discovery.TokenEndpoint, params)
}

func (c *OAuthProviderClient) RefreshTokens(ctx context.Context, refreshToken string) (*models.OAuthTokenResponse, error) {
	discovery, err := c.GetDiscovery(ctx)
	if err != nil {
		return nil, err
	}

	params := url.Values{
		"grant_type":    {"refresh_token"},
		"refresh_token": {refreshToken},
		"client_id":     {c.config.ClientID},
	}

	if c.config.ClientSecret != "" {
		params.Set("client_secret", c.config.ClientSecret)
	}

	return c.tokenRequest(ctx, discovery.TokenEndpoint, params)
}

func (c *OAuthProviderClient) IntrospectToken(ctx context.Context, token string) (*models.TokenIntrospectionResponse, error) {
	discovery, err := c.GetDiscovery(ctx)
	if err != nil {
		return nil, err
	}

	params := url.Values{"token": {token}}

	req, err := http.NewRequestWithContext(ctx, "POST", discovery.IntrospectionEndpoint, strings.NewReader(params.Encode()))
	if err != nil {
		return nil, err
	}

	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	c.applyHeaders(req)
	c.setClientAuth(req)

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("introspection failed with status %d", resp.StatusCode)
	}

	result := &models.TokenIntrospectionResponse{}
	if err := json.NewDecoder(resp.Body).Decode(result); err != nil {
		return nil, err
	}

	return result, nil
}

func (c *OAuthProviderClient) RevokeToken(ctx context.Context, token, tokenTypeHint string) error {
	discovery, err := c.GetDiscovery(ctx)
	if err != nil {
		return err
	}

	params := url.Values{"token": {token}}
	if tokenTypeHint != "" {
		params.Set("token_type_hint", tokenTypeHint)
	}

	req, err := http.NewRequestWithContext(ctx, "POST", discovery.RevocationEndpoint, strings.NewReader(params.Encode()))
	if err != nil {
		return err
	}

	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	c.applyHeaders(req)
	c.setClientAuth(req)

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	return nil
}

func (c *OAuthProviderClient) GetUserInfo(ctx context.Context, accessToken string) (*models.UserInfo, error) {
	discovery, err := c.GetDiscovery(ctx)
	if err != nil {
		return nil, err
	}

	req, err := http.NewRequestWithContext(ctx, "GET", discovery.UserInfoEndpoint, nil)
	if err != nil {
		return nil, err
	}

	c.applyHeaders(req)
	req.Header.Set("Authorization", "Bearer "+accessToken)

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("userinfo failed with status %d", resp.StatusCode)
	}

	result := &models.UserInfo{}
	if err := json.NewDecoder(resp.Body).Decode(result); err != nil {
		return nil, err
	}

	return result, nil
}

func (c *OAuthProviderClient) RequestDeviceCode(ctx context.Context, scopes []string) (*models.DeviceAuthResponse, error) {
	discovery, err := c.GetDiscovery(ctx)
	if err != nil {
		return nil, err
	}

	scope := strings.Join(c.config.Scopes, " ")
	if len(scopes) > 0 {
		scope = strings.Join(scopes, " ")
	}

	params := url.Values{
		"client_id": {c.config.ClientID},
		"scope":     {scope},
	}

	req, err := http.NewRequestWithContext(ctx, "POST", discovery.DeviceAuthorizationEndpoint, strings.NewReader(params.Encode()))
	if err != nil {
		return nil, err
	}

	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	c.applyHeaders(req)

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("device authorization failed with status %d", resp.StatusCode)
	}

	result := &models.DeviceAuthResponse{}
	if err := json.NewDecoder(resp.Body).Decode(result); err != nil {
		return nil, err
	}

	return result, nil
}

func (c *OAuthProviderClient) PollDeviceToken(ctx context.Context, deviceCode string) (*models.OAuthTokenResponse, error) {
	discovery, err := c.GetDiscovery(ctx)
	if err != nil {
		return nil, err
	}

	params := url.Values{
		"grant_type":  {"urn:ietf:params:oauth:grant-type:device_code"},
		"device_code": {deviceCode},
		"client_id":   {c.config.ClientID},
	}

	req, err := http.NewRequestWithContext(ctx, "POST", discovery.TokenEndpoint, strings.NewReader(params.Encode()))
	if err != nil {
		return nil, err
	}

	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	c.applyHeaders(req)

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		var oauthErr models.OAuthError
		if err := json.NewDecoder(resp.Body).Decode(&oauthErr); err != nil {
			return nil, fmt.Errorf("token request failed with status %d", resp.StatusCode)
		}

		switch oauthErr.Error {
		case "authorization_pending":
			return nil, ErrAuthorizationPending
		case "slow_down":
			return nil, ErrSlowDown
		case "access_denied":
			return nil, ErrAccessDenied
		case "expired_token":
			return nil, ErrExpiredToken
		default:
			return nil, fmt.Errorf("%s: %s", oauthErr.Error, oauthErr.ErrorDescription)
		}
	}

	result := &models.OAuthTokenResponse{}
	if err := json.NewDecoder(resp.Body).Decode(result); err != nil {
		return nil, err
	}

	return result, nil
}

func (c *OAuthProviderClient) ClientCredentialsGrant(ctx context.Context, scopes []string) (*models.OAuthTokenResponse, error) {
	if c.config.ClientSecret == "" {
		return nil, errors.New("client_credentials grant requires client_secret")
	}

	discovery, err := c.GetDiscovery(ctx)
	if err != nil {
		return nil, err
	}

	scope := ""
	if len(scopes) > 0 {
		scope = strings.Join(scopes, " ")
	}

	params := url.Values{
		"grant_type": {"client_credentials"},
	}
	if scope != "" {
		params.Set("scope", scope)
	}

	req, err := http.NewRequestWithContext(ctx, "POST", discovery.TokenEndpoint, strings.NewReader(params.Encode()))
	if err != nil {
		return nil, err
	}

	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	c.applyHeaders(req)
	c.setClientAuth(req)

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("client credentials grant failed with status %d", resp.StatusCode)
	}

	result := &models.OAuthTokenResponse{}
	if err := json.NewDecoder(resp.Body).Decode(result); err != nil {
		return nil, err
	}

	return result, nil
}

func (c *OAuthProviderClient) tokenRequest(ctx context.Context, endpoint string, params url.Values) (*models.OAuthTokenResponse, error) {
	req, err := http.NewRequestWithContext(ctx, "POST", endpoint, strings.NewReader(params.Encode()))
	if err != nil {
		return nil, err
	}

	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	c.applyHeaders(req)

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		var oauthErr models.OAuthError
		if err := json.NewDecoder(resp.Body).Decode(&oauthErr); err != nil {
			return nil, fmt.Errorf("token request failed with status %d", resp.StatusCode)
		}
		return nil, fmt.Errorf("%s: %s", oauthErr.Error, oauthErr.ErrorDescription)
	}

	result := &models.OAuthTokenResponse{}
	if err := json.NewDecoder(resp.Body).Decode(result); err != nil {
		return nil, err
	}

	return result, nil
}

func (c *OAuthProviderClient) setClientAuth(req *http.Request) {
	if c.config.ClientSecret != "" {
		auth := base64.StdEncoding.EncodeToString([]byte(c.config.ClientID + ":" + c.config.ClientSecret))
		req.Header.Set("Authorization", "Basic "+auth)
	}
}

func GeneratePKCE() (*models.PKCEParams, error) {
	verifier := generateRandomString(64)

	hash := sha256.Sum256([]byte(verifier))
	challenge := base64.RawURLEncoding.EncodeToString(hash[:])

	return &models.PKCEParams{
		CodeVerifier:        verifier,
		CodeChallenge:       challenge,
		CodeChallengeMethod: "S256",
	}, nil
}

func generateRandomString(length int) string {
	const charset = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-._~"
	b := make([]byte, length)
	rand.Read(b)
	for i := range b {
		b[i] = charset[b[i]%byte(len(charset))]
	}
	return string(b)
}

File: packages/client-sdk/src/oauth-provider-client.ts
/**
 * OAuth Provider Client
 *
 * Utility for third-party applications integrating with Auth Gateway as an OAuth/OIDC provider.
 * Handles authorization code flow, PKCE, token exchange, and OIDC discovery.
 *
 * @example
 * ```typescript
 * import { OAuthProviderClient } from '@auth-gateway/client-sdk/oauth-provider';
 *
 * const client = new OAuthProviderClient({
 *   issuer: 'https://auth.example.com',
 *   clientId: 'your-client-id',
 *   clientSecret: 'your-secret', // Optional for public clients
 *   redirectUri: 'https://yourapp.com/callback',
 *   scopes: ['openid', 'profile', 'email'],
 *   usePKCE: true, // Recommended
 * });
 *
 * // Get authorization URL
 * const { url, state, codeVerifier } = await client.getAuthorizationUrl();
 * window.location.href = url;
 *
 * // Handle callback (after redirect)
 * const tokens = await client.exchangeCode(code, codeVerifier);
 *
 * // Get user info
 * const userInfo = await client.getUserInfo(tokens.access_token);
 * ```
 */

import type {
  OIDCDiscoveryDocument,
  JWKS,
  TokenResponse,
  TokenIntrospectionResponse,
  UserInfoResponse,
  DeviceAuthResponse,
  IDTokenClaims,
  PKCEParams,
  AuthorizationUrlOptions,
  AuthorizationUrlResult,
} from './types/oauth-provider';

export interface OAuthProviderClientConfig {
  issuer: string;
  clientId: string;
  clientSecret?: string;
  redirectUri: string;
  scopes?: string[];
  usePKCE?: boolean;
}

export class OAuthProviderClient {
  private config: OAuthProviderClientConfig;
  private discovery?: OIDCDiscoveryDocument;
  private jwks?: JWKS;

  constructor(config: OAuthProviderClientConfig) {
    this.config = {
      usePKCE: true,
      scopes: ['openid'],
      ...config,
    };
  }

  async getDiscovery(): Promise<OIDCDiscoveryDocument> {
    if (this.discovery) return this.discovery;

    const response = await fetch(
      `${this.config.issuer}/.well-known/openid-configuration`
    );

    if (!response.ok) {
      throw new Error(`Failed to fetch discovery document: ${response.status}`);
    }

    this.discovery = await response.json();
    return this.discovery!;
  }

  async getJWKS(): Promise<JWKS> {
    if (this.jwks) return this.jwks;

    const discovery = await this.getDiscovery();
    const response = await fetch(discovery.jwks_uri);

    if (!response.ok) {
      throw new Error(`Failed to fetch JWKS: ${response.status}`);
    }

    this.jwks = await response.json();
    return this.jwks!;
  }

  async getAuthorizationUrl(options?: AuthorizationUrlOptions): Promise<AuthorizationUrlResult> {
    const discovery = await this.getDiscovery();

    const state = options?.state || this.generateRandomString(32);
    const nonce = options?.nonce || this.generateRandomString(32);

    const params = new URLSearchParams({
      response_type: 'code',
      client_id: this.config.clientId,
      redirect_uri: this.config.redirectUri,
      scope: options?.scope || this.config.scopes!.join(' '),
      state,
      nonce,
    });

    if (options?.prompt) {
      params.set('prompt', options.prompt);
    }
    if (options?.login_hint) {
      params.set('login_hint', options.login_hint);
    }

    let codeVerifier: string | undefined;

    if (this.config.usePKCE) {
      const pkce = await this.generatePKCE();
      codeVerifier = pkce.code_verifier;
      params.set('code_challenge', pkce.code_challenge);
      params.set('code_challenge_method', pkce.code_challenge_method);
    }

    return {
      url: `${discovery.authorization_endpoint}?${params.toString()}`,
      state,
      nonce,
      codeVerifier,
    };
  }

  async exchangeCode(code: string, codeVerifier?: string): Promise<TokenResponse> {
    const discovery = await this.getDiscovery();

    const params = new URLSearchParams({
      grant_type: 'authorization_code',
      code,
      redirect_uri: this.config.redirectUri,
      client_id: this.config.clientId,
    });

    if (this.config.clientSecret) {
      params.set('client_secret', this.config.clientSecret);
    }

    if (codeVerifier) {
      params.set('code_verifier', codeVerifier);
    }

    const response = await fetch(discovery.token_endpoint, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
      },
      body: params.toString(),
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.error_description || error.error || 'Token exchange failed');
    }

    return response.json();
  }

  async refreshTokens(refreshToken: string): Promise<TokenResponse> {
    const discovery = await this.getDiscovery();

    const params = new URLSearchParams({
      grant_type: 'refresh_token',
      refresh_token: refreshToken,
      client_id: this.config.clientId,
    });

    if (this.config.clientSecret) {
      params.set('client_secret', this.config.clientSecret);
    }

    const response = await fetch(discovery.token_endpoint, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
      },
      body: params.toString(),
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.error_description || error.error || 'Token refresh failed');
    }

    return response.json();
  }

  async introspectToken(token: string): Promise<TokenIntrospectionResponse> {
    const discovery = await this.getDiscovery();

    const params = new URLSearchParams({ token });

    const headers: Record<string, string> = {
      'Content-Type': 'application/x-www-form-urlencoded',
    };

    if (this.config.clientSecret) {
      headers['Authorization'] = `Basic ${this.base64Encode(`${this.config.clientId}:${this.config.clientSecret}`)}`;
    }

    const response = await fetch(discovery.introspection_endpoint, {
      method: 'POST',
      headers,
      body: params.toString(),
    });

    if (!response.ok) {
      throw new Error('Token introspection failed');
    }

    return response.json();
  }

  async revokeToken(token: string, tokenTypeHint?: 'access_token' | 'refresh_token'): Promise<void> {
    const discovery = await this.getDiscovery();

    const params = new URLSearchParams({ token });
    if (tokenTypeHint) {
      params.set('token_type_hint', tokenTypeHint);
    }

    const headers: Record<string, string> = {
      'Content-Type': 'application/x-www-form-urlencoded',
    };

    if (this.config.clientSecret) {
      headers['Authorization'] = `Basic ${this.base64Encode(`${this.config.clientId}:${this.config.clientSecret}`)}`;
    }

    await fetch(discovery.revocation_endpoint, {
      method: 'POST',
      headers,
      body: params.toString(),
    });
  }

  async getUserInfo(accessToken: string): Promise<UserInfoResponse> {
    const discovery = await this.getDiscovery();

    const response = await fetch(discovery.userinfo_endpoint, {
      headers: {
        'Authorization': `Bearer ${accessToken}`,
      },
    });

    if (!response.ok) {
      throw new Error('Failed to fetch user info');
    }

    return response.json();
  }

  decodeIdToken(idToken: string): IDTokenClaims {
    const parts = idToken.split('.');
    if (parts.length !== 3) {
      throw new Error('Invalid ID token format');
    }

    const payload = JSON.parse(this.base64Decode(parts[1] as string));

    if (payload.iss !== this.config.issuer) {
      throw new Error('Invalid issuer');
    }
    if (payload.aud !== this.config.clientId) {
      throw new Error('Invalid audience');
    }
    if (payload.exp < Date.now() / 1000) {
      throw new Error('Token expired');
    }

    return payload as IDTokenClaims;
  }

  async requestDeviceCode(scopes?: string[]): Promise<DeviceAuthResponse> {
    const discovery = await this.getDiscovery();

    const params = new URLSearchParams({
      client_id: this.config.clientId,
      scope: scopes?.join(' ') || this.config.scopes!.join(' '),
    });

    const response = await fetch(discovery.device_authorization_endpoint, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
      },
      body: params.toString(),
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.error_description || error.error || 'Device authorization failed');
    }

    return response.json();
  }

  async pollDeviceToken(deviceCode: string): Promise<TokenResponse> {
    const discovery = await this.getDiscovery();

    const params = new URLSearchParams({
      grant_type: 'urn:ietf:params:oauth:grant-type:device_code',
      device_code: deviceCode,
      client_id: this.config.clientId,
    });

    const response = await fetch(discovery.token_endpoint, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
      },
      body: params.toString(),
    });

    if (!response.ok) {
      const error = await response.json();
      if (error.error === 'authorization_pending' || error.error === 'slow_down') {
        throw new DeviceFlowPendingError(error.error);
      }
      throw new Error(error.error_description || error.error || 'Device token request failed');
    }

    return response.json();
  }

  async clientCredentialsGrant(scopes?: string[]): Promise<TokenResponse> {
    if (!this.config.clientSecret) {
      throw new Error('Client credentials grant requires client_secret');
    }

    const discovery = await this.getDiscovery();

    const params = new URLSearchParams({
      grant_type: 'client_credentials',
      scope: scopes?.join(' ') || '',
    });

    const response = await fetch(discovery.token_endpoint, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
        'Authorization': `Basic ${this.base64Encode(`${this.config.clientId}:${this.config.clientSecret}`)}`,
      },
      body: params.toString(),
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.error_description || error.error || 'Client credentials grant failed');
    }

    return response.json();
  }

  async generatePKCE(): Promise<PKCEParams> {
    const codeVerifier = this.generateRandomString(64);
    const encoder = new TextEncoder();
    const data = encoder.encode(codeVerifier);
    const hash = await crypto.subtle.digest('SHA-256', data);
    const codeChallenge = this.base64UrlEncode(new Uint8Array(hash));

    return {
      code_verifier: codeVerifier,
      code_challenge: codeChallenge,
      code_challenge_method: 'S256',
    };
  }

  private generateRandomString(length: number): string {
    const charset = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-._~';
    const randomValues = crypto.getRandomValues(new Uint8Array(length));
    return Array.from(randomValues)
      .map(x => charset[x % charset.length])
      .join('');
  }

  private base64UrlEncode(buffer: Uint8Array): string {
    return this.base64Encode(String.fromCharCode(...buffer))
      .replace(/\+/g, '-')
      .replace(/\//g, '_')
      .replace(/=+$/, '');
  }

  private base64Encode(str: string): string {
    if (typeof btoa !== 'undefined') {
      return btoa(str);
    }
    return Buffer.from(str).toString('base64');
  }

  private base64Decode(str: string): string {
    if (typeof atob !== 'undefined') {
      return atob(str);
    }
    return Buffer.from(str, 'base64').toString('utf-8');
  }
}

export class DeviceFlowPendingError extends Error {
  constructor(public code: 'authorization_pending' | 'slow_down') {
    super(`Device flow ${code}`);
    this.name = 'DeviceFlowPendingError';
  }
}
